columnflow/tasks/reduction.py

```diff
@@ -102,11 +102,27 @@ def run(self):
         # get shift dependent aliases
         aliases = self.shift_inst.x("column_aliases_selection_dependent", {})
 
-        # define nano columns that should be kept, and that need to be loaded
-        keep_columns = set(self.config_inst.x.keep_columns.get(self.task_family, ["*"]))
-        load_columns = keep_columns | set(mandatory_coffea_columns)
-        load_columns_nano = [Route(column).nano_column for column in load_columns]
-        route_filter = RouteFilter(keep_columns)
+        # define columns that will be written
+        write_columns = set(self.config_inst.x.keep_columns.get(self.task_family, ["*"]))
+        route_filter = RouteFilter(write_columns)
+
+        # define columns that need to be read
+        read_columns = write_columns | set(mandatory_coffea_columns) | set(aliases.values())
+        read_columns = {Route(c) for c in read_columns}
+
+        # define columns to read for the differently structured selection masks
+        read_sel_columns = {Route("steps.*")}
+        # add event routes
+        read_sel_columns |= {("event" + r) for r in read_columns}
+        # add object masks, depending on the columns to write
+        # (as object masks are dynamic and deeply nested, preload the meta info to access fields)
+        masks_meta = inputs["selection"]["results"].load(formatter="dask_awkward").objects
+        write_columns_toplevel = {Route(r)[0] for r in write_columns}
+        for src_field in masks_meta.fields:
+            for dst_field in masks_meta[src_field].fields:
+                if law.util.multi_match(dst_field, write_columns_toplevel):
+                    read_sel_columns.add(Route(f"objects.{src_field}.{dst_field}"))
+        del masks_meta
```

Comment:
Awesome, thanks a lot for implementing this! Just to make sure I understand: There is no "selection" of fields that are produced with selectors, right? I would have somehow guessed that those might be part of the `uses` set of the producers, though that would probably end up in the `read_columns` in that case. I think the way it's setup right now, there isn't a real connection between producers and selectors, right? (I think that is something I stumbled upon the last time I reviewed something as well). 

As discussed then, this is fine. It's still great to have specific mention of the selector fields though because
- it's clearly readable and understandable
- it's easier to manuplate the selector columns if need be
:+1:

Reply:
> There is no "selection" of fields that are produced with selectors, right? I would have somehow guessed that those might be part of the uses set of the producers, though that would probably end up in the read_columns in that case.

In fact there was a small typo in there that I fixed now (removed lines 115-116 in comment above). The selectors produce two files: one with selection steps (`steps.*`) and per-object masks (`objects.*`, e.g. `objects.Jet.Jet`), and one with additional columns *if the selectors creates some*.

The `read_sel_columns` only selects the masks in the first file. The event columns are opened with the usual `read_columns` as before.

Reply:
Thanks for clarifying! One final question: Do we really need to load `staps.*` in the reduction task? I would have assumed we only need this information for the cutflow task(s), which are separated from this task here iirc, right?